### PR TITLE
RFC: Rename ext to pipes

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

We are understandably getting feedback that `ext` is a confusing name. With it becoming increasingly likely that we will be calling the successor to source and observable source assets "external assets" the situation compounds.

This PR proposes a new name: Dagster Pipes.

Pipes stands for:

* **P**rotocol for 
* **I**nter
* **P**rocess 
* **E**xecution 
* **S**treaming logs and metadata

I think this is memorable and accurately describes the intent of the system, as well as providing a metaphor for a broader set of stakeholders. "This is the pipe that connects Dagster to Databricks." seems like a reasonable thing to say.

I also like the way the entire taxonomy plays out here.

There are "piped process". Those are the processes that run externally. To invoke them you use classes like `PipedK8sPod`. "This launches a piped k8s pod" rolls off the tongue. In the external script you write `init_dagster_piped_process` which returns a `PipedProcessContext`. (We could also name `PipedProcessContext` `PipedProcessInvocation` if we are having context fatigue)

For the protocol context manager I propose the following convention:

```python
# pipes_protocol returns PipedClientRequest
with pipes_protocol(
    context,
    EnvPipedContextInjector(),
    reader,
    extras={"storage_root": "/tmp/"},
) as piped_client_req: ...
```

"Pipeable" is also a useful word to describe the serializable data. For example we have `PipeablePartitionKeyRange` and `PipeableDataProvenance`. I think this is fairly clear and demonstrated that these classes could go in either direction.

Lastly there is `Pipes` which is used to describe the overall system. E.g. `DagsterPipesError`

As a bonus `dagster-pipes` is also not occupied by a squatter.

I need some help getting BK up and running, but other than that I think this PR is good to go with light modifications.

## How I Tested These Changes

BK

### CODE REVIEW

**Comment by schrockn** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16749** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16749" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16749?utm_source=stack-comment).

**Comment by petehunt** [comment]
I like this a lot

**Comment by smackesey** [comment]
I also like this.

We did discuss "pipes" in our initial naming sessions (alongside e.g. "relay") and I forget why exactly we rejected it-- probably bc it already has strong associations in an IPC context. I do think that's a slight issue with e.g. `PipedSubprocess`, since a `PipedSubprocess` is literally not using pipes by default for message comms. However that's nbd and I think it's an extremely useful metaphor that will help users intuitively grasp the purpose of the protocol wrt to external systems.

**Comment by sryza** [comment]
I like it

**Comment by PedramNavid** [comment]
My main concern here is that it starts to sound a lot like unix pipes. Just hearing “dagster pipes” doesn’t invoke much for me. Is it fair to think of this as an RPC?

File: python_modules/dagster/dagster/_core/pipes/client.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+class PipedContextInjector(ABC):
+    @abstractmethod
+    @contextmanager
+    def inject_context(
+        self, context_data: "PipedProcessContextData"
+    ) -> Iterator[PipedProcessParams]: ...
+
+
+class PipedMessageReader(ABC):
```
[the remark this answers is not in the document]
Comment (PedramNavid) [reply python_modules/dagster/dagster/_core/pipes/client.py:35]: I would agree with @alangenfeld, I read PipedMessageReader and PipedContextInjector as message readers and context injectors that are piped.

[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/_core/pipes/client.py:35]: Cool between the two of you convinced that we should try alternative nomenclature for those

**Comment by schrockn** [comment]
Doesn't sound like there are any objections to this, and general excitement. We are going to proceed with this name change

**Comment by schrockn** [comment]
Landed in #16756

**Pull request closed without merging**